core/types/transactions/payloads.go

```diff
@@ -166,12 +169,175 @@ func (a *ActionExecution) Type() PayloadType {
 type ActionCall struct {
 	DBID      string
 	Action    string
-	Arguments []string
+	Arguments []*EncodedValue
 	// NilArg indicates for each of the elements in Arguments if the value is
 	// nil rather than just empty.
 	NilArg []bool `rlp:"optional"`
 }
 
+// EncodedValue is used to encode a value with its type specified
+type EncodedValue struct {
+	Type DataType
+	// The double slice handles arrays of encoded values.
+	// If there is only one element, the outer slice will have length 1.
+	Data [][]byte `rlp:"optional"`
+}
+
+// Decode decodes the encoded value to its native Go type.
+func (e *EncodedValue) Decode() (any, error) {
+	// decodeScalar decodes a scalar value from a byte slice.
+	decodeScalar := func(data []byte, t *types.DataType) (any, error) {
+		if data == nil {
+			return nil, nil
+		}
+		switch t.Name {
+		case types.TextType.Name:
+			return string(data), nil
+		case types.IntType.Name:
+			if len(data) != 8 {
+				return nil, fmt.Errorf("int must be 8 bytes")
+			}
+			return int64(binary.BigEndian.Uint64(data)), nil
+		case types.BlobType.Name:
+			return data, nil
+		case types.UUIDType.Name:
+			if len(data) != 16 {
+				return nil, fmt.Errorf("uuid must be 16 bytes")
+			}
+			var uuid types.UUID
+			copy(uuid[:], data)
+			return uuid, nil
+		case types.BoolType.Name:
+			return data[0] == 1, nil
+		case types.NullType.Name:
+			return nil, nil
+		default:
+			return nil, fmt.Errorf("cannot decode type %s", t.Name)
+		}
+	}
+
+	// check if it is an array
+	if e.Type.IsArray {
+		decoded := make([]any, 0)
+		for _, elem := range e.Data {
+			dec, err := decodeScalar(elem, e.Type.toTypes())
```

Comment:
Probably spare some conversions and allocations and just pass `e.Type.Name` to `decodeScalar` because that's all it needs.  Array context is already established here.